### PR TITLE
feat(clipper): Implement concurrent file reading and refactor project structure

### LINKED ISSUE #23: Explore Concurrency (goroutines and channels) for Copying from Multiple Files
When copying content from multiple files, clipper currently processes them sequentially. This might be inefficient, especially when dealing with a large number of files.

We should investigate the use of concurrency with [goroutines](https://go.dev/doc/effective_go#goroutines) and channels to parallelize the reading of content from multiple files. This could significantly speed up the copying process in such scenarios.

**See:**
- [How to read multiple files...](https://www.reddit.com/r/golang/comments/18zaofb/how_to_read_multiple_files_and_close_them/)

**Tasks:**

1. Design and implement a concurrent solution using goroutines and channels to read content from multiple files simultaneously.
2. Consider potential synchronization issues and error handling in a concurrent environment.
3. Benchmark the performance improvement when copying content from multiple files.
4. Update documentation to explain the use of concurrency.

### PR DESCRIPTION
**Pull request by supitsdu**
This PR enhances Clipper by introducing concurrent file reading using goroutines and channels in the new `cli/reader/reader.go` package. It refactors `cli/clipper/clipper.go` to utilize these improvements, improving performance and maintainability. The test structure is also reorganized for clarity.

### Changes

- **New `cli/reader/reader.go` Package:**
  - Implements concurrent file reading with goroutines and channels.
  - Introduces `ReadContentConcurrently` function for efficient file handling.

- **Refactored `cli/clipper/clipper.go`:**
  - Updated to leverage `cli/reader/reader.go` for file reading operations.
  - Enhances code clarity and separation of concerns.

- **Enhanced Test Structure:**
  - Introduces `tests/tests.go` for shared testing utilities.
  - Restructures tests into `tests/clipper/clipper_test.go` and `tests/reader/reader_test.go`.

### Tasks

- [x] Implement concurrent file reading with `cli/reader/reader.go`.
- [x] Refactor `cli/clipper/clipper.go` for improved readability and performance.
- [x] Organize tests into dedicated files for better maintainability and focus.
- [x] Benchmark performance improvements with concurrent file reading.
- [x] Document benchmark results and optimizations for future reference.

### Next Steps

Upon merging, focus will shift to benchmarking the application to quantify performance gains from concurrent file reading. This will inform future optimizations and enhancements to Clipper's file handling capabilities.

Closes #23

### CODE REVIEW

File: cli/reader/reader.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	input, err := io.ReadAll(os.Stdin)
+	if err != nil {
+		return "", fmt.Errorf("error reading from stdin: %w", err)
+	}
+	return string(input), nil
+}
+
+// ReadContentConcurrently reads content from multiple readers concurrently and returns the results.
+// This function utilizes goroutines to perform concurrent reads, improving performance for multiple files.
+func ReadContentConcurrently(readers []ContentReader) ([]string, error) {
```
Comment (ccoVeille) [inline cli/reader/reader.go:46]: Maybe a variadic here


```suggestion
func ReadContentConcurrently(readers ...ContentReader) ([]string, error) {
```

This way calling the function would be more readable

File: cli/reader/reader.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	var sb strings.Builder
+	for _, content := range results {
+		sb.WriteString(content + "\n")
+	}
+	return sb.String()
+}
+
+// ParseContent aggregates content from the provided readers, or returns the direct text if provided.
+// This function first checks for direct text input, then reads from the provided readers concurrently.
+func ParseContent(directText *string, readers ...ContentReader) (string, error) {
```
Comment (ccoVeille) [inline cli/reader/reader.go:89]: You could have refactored the*string in this PR :)

File: cli/reader/reader.go (hunk 1 above)
Comment (ccoVeille) [reply cli/reader/reader.go:46]: But because of getReaders, it may be pointless

Comment (supitsdu) [reply cli/reader/reader.go:46]: > Maybe a variadic here
> 
> This way calling the function would be more readable

I'll work on some refactoring next, which tackles the #30 issue, when I open a PR I may include your suggestion then. Thanks, as always.

File: cli/reader/reader.go (hunk 2 above)
Comment (supitsdu) [reply cli/reader/reader.go:89]: I was going to... but I don't quite appreciate putting too many changes in a single PR (It just doesn't look right to me). 

I'm still learning tho, later I might change this mindset.

I'll work on the #30 issue as mentioned [here](https://github.com/supitsdu/clipper/pull/34#discussion_r1660232112)

File: cli/reader/reader.go (hunk 1 above)
Comment (supitsdu) [reply cli/reader/reader.go:46]: See PR #37

**Pull request merged**